src/DataGrid.tsx

```diff
@@ -700,8 +701,21 @@ function DataGrid<R, SR>({
     if (!current) return;
 
     if (typeof idx === 'number' && idx > lastFrozenColumnIndex) {
+      rowIdx ??= selectedPosition.rowIdx;
+      if (!isCellWithinBounds({ rowIdx, idx })) return;
       const { clientWidth } = current;
-      const { left, width } = columnMetrics.get(columns[idx])!;
+      const column = columns[idx];
+      const selectedColumnMetrics = columnMetrics.get(column)!;
+      let { width } = selectedColumnMetrics;
+      const { left } = selectedColumnMetrics;
+      const row = rows[rowIdx];
+      if (!isGroupRow(row)) {
+        const colSpan = getColSpan(column, lastFrozenColumnIndex, { type: 'ROW', row });
+        if (colSpan !== undefined) {
+          const colSpanColumnMetrics = columnMetrics.get(columns[column.idx + colSpan - 1])!;
+          width = colSpanColumnMetrics.left - left + colSpanColumnMetrics.width;
+        }
+      }
       const isCellAtLeftBoundary = left < scrollLeft + totalFrozenColumnWidth;
       const isCellAtRightBoundary = left + width > clientWidth + scrollLeft;
```

Comment:
Since we only care about `left` or `right`, we could do something like
```ts
let right;
if (colSpan === undefined) {
  right = left + width;
} else {
  const { left, width } = columnMetrics.get(columns[column.idx + colSpan - 1])!;
  right = left + width;
}
// ...
const isCellAtRightBoundary = right > clientWidth + scrollLeft;
```
Should be easier to understand than mixing starting column left with last column left.